### ISSUE TITLE
Popover with `data-trigger="focus"` does not work on iOS Safari when used on <a> without `class="btn"`

### ISSUE THREAD

**Issue by MrOrz** [twbs/bootstrap#15935]
After #11788 and #14156 the popover with `data-trigger="focus"` is working very well. However, there is a super weird bug that only happens in iOS safari.

Consider the following 2 `<a>`, the only difference between them is `class="btn"`.

```html
  <a tabindex="0" role="button" data-toggle="popover" data-trigger="focus"
      class="btn"
      data-content="...">
    Popover anchor w/ class 'btn'
  </a>

  <a tabindex="1" role="button" data-toggle="popover" data-trigger="focus"
      data-content="...">
    Popover anchor w/o class 'btn'
  </a>
```

The problem is, popover works fine for the first `<a>`, but not the second.

Hope the recording below will help better illustrating the problem:

![1](https://cloud.githubusercontent.com/assets/108608/6390690/6134dac0-bde9-11e4-9b62-8bd5282df2d8.gif)

Demo Plunker: http://embed.plnkr.co/M0NIAPtXOet8ppl2dShP/preview
(Please open using iPhone or iOS simulator. OSX Chrome, Android Chrome, OSX Safari, OSX Firefox works perfectly on both `<a>`s.)

---

It seems that the focus event is somehow not triggered in iOS Safari. When I put a breakpoint in the focus event callback, the first `<a>` invokes the callback as expected while the second `<a>` does not.

![2](https://cloud.githubusercontent.com/assets/108608/6390758/ec8cbda4-bde9-11e4-9f04-a713736772ef.gif)

No idea why this happens ._.

**Comment by twbs-lmvtfy** [comment]
Hi @MrOrz!

You appear to have posted a live example (http://run.plnkr.co/plunks/M0NIAPtXOet8ppl2dShP/), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* [W002](https://github.com/twbs/bootlint/wiki/W002): `<head>` is missing X-UA-Compatible `<meta>` tag that disables old IE compatibility modes

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by MrOrz** [comment]
`X-UA-Compatible` added in example and saved :)

The url is the same as before: http://run.plnkr.co/plunks/M0NIAPtXOet8ppl2dShP/

**Comment by cvrebert** [comment]
* What about iOS 8.1?
* Does it work if you just add `href="#"` ?
* Does it work if you just add `style="cursor: pointer"` ?

**Comment by cvrebert** [comment]
Probable X-Ref: #15277

**Comment by cvrebert** [comment]
Confirmed on iOS 8.1.
Without `.btn`, you need the `href="#"`. `.btn` doesn't strictly need the href on account of its own `cursor: pointer`.
As not using `.btn` is legitimate, I'm just going to add `href="#"` for the sake of genericity & simplicity, effectively reverting #15277, which was only a very minor nice-to-have anyway.

> It seems that the focus event is somehow not triggered in iOS Safari

Yes, iOS Safari is stingy/quirky about elements being clickable/focusable; see https://developer.mozilla.org/en-US/docs/Web/Events/click#Safari_Mobile

**Comment by patrickhlauke** [comment]
If at all possible, the best solution is to use an actual ``<button>`` element...but yes, failing that, explicitly adding a ``cursor:pointer`` would be the cleanest way of doing it, as adding an ``href="#"`` will then also require you to ``preventDefault()`` the action of the link.

@cvrebert perhaps adding a rule to the CSS along lines of ``*[role=button] { cursor: pointer; }`` may be a nice touch?

**Comment by cvrebert** [comment]
@patrickhlauke `<button>` isn't an option because of fucking Safari: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Button#Clicking_and_focus
Way ahead of you on that last thing: https://github.com/necolas/normalize.css/pull/379

**Comment by patrickhlauke** [comment]
ah, safari...gotta love it. regarding normalize, as that's been sitting there since october, worth doing our own patch?

**Comment by cvrebert** [comment]
Sure, go for it. Let's see what mdo thinks.

**Comment by patrickhlauke** [comment]
What would be best? making a change to our local normalize.less, or adding the above to, say, buttons.less ?

**Comment by cvrebert** [comment]
Our normalize.less is basically vendored and thus shouldn't be modified.

**Comment by cvrebert** [comment]
scaffolding.less might be more appropriate than buttons.less

**Comment by XhmikosR** [comment]
@patrickhlauke @cvrebert: Maybe we should add a comment that an upstream patch has been submitted so that we revise this and remove it if added in normalize.css?

**Comment by cvrebert** [comment]
@XhmikosR Agreed. Might wanna move your comment to the PR.

**Comment by XhmikosR** [comment]
@patrickhlauke: you should not merge that yet. https://github.com/twbs/bootstrap/issues/15935#issuecomment-76694397

**Comment by patrickhlauke** [comment]
Well I realise that NOW...so, how do I undo this?

**Comment by XhmikosR** [comment]
Just add a new commit with the comment.

Well, I didn't have the time to move the comment to the PR. The problem is that there are so many PRs and issues that things like this happen. Just look at the notifications more closely.

**Comment by patrickhlauke** [comment]
Done. Will be more careful in future (maybe bad idea to have my notifications set to "firehose"...)

**Comment by DavidVII** [comment]
I want to add that this bug is also happening for me on Safari 8.0.4. Not just on iOS. I resolved this by adding the `tabindex` attribute the `.btn` element. Safari is having some weird issues trying to trigger the focus event it seems.

**Comment by cvrebert** [comment]
@DavidVII That's a very different problem. As already mentioned in the docs, `tabindex` is always required for dismiss-on-next-click popovers.

**Comment by DavidVII** [comment]
@cvrebert My mistake. Figured since it worked on all other browsers except safari that this was related some how.

**Comment by cvrebert** [comment]
Though to be fair, Safari is the main reason that the tabindex is necessary; see https://gist.github.com/cvrebert/68659d0333a578d75372
